### PR TITLE
MAKE-455 Add in-memory buffered sender

### PR DESCRIPTION
**Pull request by Kimchelly**

### CODE REVIEW

File: send/inmemory.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	s.reset = func() {
+		fallback.SetPrefix(fmt.Sprintf("[%s] ", s.Name()))
+	}
+
+	return s, nil
+}
+
+// Get returns at most n most recent messages. If there are fewer than n messages, returns all currently
+// available messages in the buffer.
+func (s *InMemorySender) Get(n int) ([]message.Composer, error) {
```
Comment (tychoish) [inline send/inmemory.go]: can we have methods (or helpers?) that convert these to strings or slices of interfaces (e.g. the output of .Raw()) for easier use?

File: send/inmemory.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+	if n > len(s.buffer) {
+		return append(make([]message.Composer, 0, n), s.buffer[:s.head]...), nil
+	}
+
+	start := (s.head - n)
+	if start < 0 {
+		start = start + len(s.buffer)
+		tmp := append(make([]message.Composer, 0, n), s.buffer[start:len(s.buffer)]...)
+		return append(tmp, s.buffer[:s.head]...), nil
```
Comment (tychoish) [inline send/inmemory.go]: this is

File: send/inmemory.go (hunk 1 above)
Comment (Kimchelly) [reply send/inmemory.go]: I added a Raw(n) and String(n) method to the sender. String returns all the messages as one string, Raw returns the Composer.Raw output.

File: send/inmemory.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"log"
+	"os"
+	"sync"
+
+	"github.com/mongodb/grip/message"
+)
+
+// InMemorySender represents an in-memory buffered sender with a fixed message capacity.
+type InMemorySender struct {
+	*Base
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply send/inmemory.go]: Done.

**Comment by Kimchelly** [comment]
I rewrote and added some tests for the correctness of Get because I found a bug.

File: send/send_test.go [hunk 4]
```diff
@@ -172,6 +172,10 @@ func (s *SenderSuite) SetupTest() {
 		ref:  "master",
 	}
 	s.NoError(s.senders["gh-status-mocked"].SetFormatter(MakeDefaultFormatter()))
+
+	s.senders["inmemory"], err = NewInMemorySender("inmemory", l, 100)
+	s.Require().NoError(err)
+	s.NoError(s.senders["inmemory"].SetFormatter(MakeDefaultFormatter()))
```
Comment (tychoish) [inline send/send_test.go]: maybe adds copies of this with much smaller and much larger caps just to make sure that works as expectd with different sizes.

Comment (Kimchelly) [reply send/send_test.go]: Done. Should this also apply to the inmemory-specific tests or should I just leave those as is?

**Pull request merged**